src/schedule/zephyr_dma_domain.c

```diff
@@ -130,57 +141,92 @@ static void zephyr_dma_domain_thread_fn(void *p1, void *p2, void *p3)
 
 static void dma_irq_handler(void *data)
 {
-	struct zephyr_dma_domain_data *zephyr_dma_domain_data;
+	struct zephyr_dma_domain_irq *irq_data;
 	struct zephyr_dma_domain_thread *dt;
-	struct dma_chan_data *channel;
 	struct k_sem *sem;
-	int channel_index, irq;
+	struct list_item *i;
+	struct zephyr_dma_domain_channel *chan_data;
 
-	zephyr_dma_domain_data = data;
-	channel = zephyr_dma_domain_data->channel;
-	channel_index = channel->index;
-	sem = &zephyr_dma_domain_data->dt->sem;
-	irq = zephyr_dma_domain_data->irq;
-	dt = zephyr_dma_domain_data->dt;
+	irq_data = data;
+	sem = &irq_data->dt->sem;
+	dt = irq_data->dt;
+
+	list_for_item(i, &irq_data->channels) {
+		chan_data = container_of(i, struct zephyr_dma_domain_channel, list);
+
+		if (dma_interrupt_legacy(chan_data->channel, DMA_IRQ_STATUS_GET))
+			dma_interrupt_legacy(chan_data->channel, DMA_IRQ_CLEAR);
+	}
 
 	/* clear IRQ */
-	dma_interrupt_legacy(channel, DMA_IRQ_CLEAR);
-	interrupt_clear_mask(irq, BIT(channel_index));
+	interrupt_clear_mask(irq_data->intid, BIT(chan_data->channel->index));
```

Comment:
Agreed, this isn't right. Will probably end up passing 0 as the mask argument to interrupt_clear_mask since it doesn't seem to be used anyways. Otherwise, `interrupt_clear_mask` should have been called for each channel when clearing interrupts.